### PR TITLE
Feature/add caching layer

### PR DESCRIPTION
**Pull request by nnevalainen**
Add service layer to abstract caching. Cache getAllUsers and getPredictions

### CODE REVIEW

File: src/bubbleBurster.ts [hunk 1]
```diff
@@ -20,28 +24,36 @@ class BubbleBurster {
...
-    const userGroupAgg: UserGroupAggregation = await this.elastic.searchQuery(searchQuery, maxResults, offices, roles);
-    const userBuckets: UserBucket[] = userGroupAgg.buckets;
-    await this.getAllUsers();
-
-    const results: UserResult[] = this.buildResult(userBuckets);
-    return results;
+    const hash = generateHash(searchQuery + maxResults.toString() + offices.toString() + roles.toString()).toString();
+    const key = `predictions_${hash}`;
+    return this.cache.get(key, async () => {
+      const userGroupAgg: UserGroupAggregation = await this.elastic.searchQuery(
```
Comment (ArktinenSieni) [inline src/bubbleBurster.ts]: Lines 30-39 could be separated to its own function for clarity. To be named as `getUsersStoreFunction()` or something like that, similar to the Cache-wrapper.

File: src/utils/cache-service.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
+import NodeCache from 'node-cache';
+
```
Comment (ArktinenSieni) [inline src/utils/cache-service.ts]: I'm not sure how `node-cache` is intented to be used, but this seems like to be a general cache for everything? Would it make any sense to have typed cache-wrapper, to ensure the return type of the `get` function?

Something like `export class Cache<T> {...`
and then having `async get(key: string, storeFunction: () => T): T {...`

File: src/bubbleBurster.ts (hunk 1 above)
Comment (nnevalainen) [reply src/bubbleBurster.ts]: Sort all input values before generating key

File: src/bubbleBurster.ts [hunk 3]
```diff
@@ -9,9 +11,11 @@ class BubbleBurster {
   @Inject
   private elastic!: Elastic;
   private allUsers: AllUsers;
+  private cache: Cache;
 
   constructor() {
     this.allUsers = {};
+    this.cache = new Cache(60 * 60 * 24); // cache for 24 hours
```
Comment (nnevalainen) [inline src/bubbleBurster.ts]: Get this from static file

File: src/bubbleBurster.ts [hunk 4]
```diff
@@ -20,28 +24,36 @@ class BubbleBurster {
...
+      try {
         const rawUsers = await this.elastic.getAllUsers();
         const usersArr = this.formatUsers(rawUsers);
-        // Create object of user array
-        this.allUsers = usersArr.reduce((obj: any, user: CompactUser) => ({ ...obj, [user.login]: user }), {});
+        return usersArr.reduce((allUsers: AllUsers, user: CompactUser) => ({ ...allUsers, [user.login]: user }), {});
+      } catch (error) {
+        console.error(`getAllUsers: unexpected error, returning empty list`);
+        console.error(`getAllUsers: ${error}`);
+        return {};
```
Comment (nnevalainen) [inline src/bubbleBurster.ts]: Throw an error

File: src/utils/cache-service.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+  }
+
+  delete(keys: string[]) {
+    this.cache.del(keys);
+  }
+
+  flush() {
+    this.cache.flushAll();
+  }
+}
```
Comment (nnevalainen) [inline src/utils/cache-service.ts]: Don't store errors into cache. For example storeFunction could throw exception and this catch it

Comment (nnevalainen) [reply src/utils/cache-service.ts]: Rename storeFunction to callback

**Pull request merged**